File: kata-python/book_python_algorithm_interview/ch6/819_most_common_word.py
# https://leetcode.com/problems/most-common-word

import re
import collections

paragraph = "Bob hit a ball, the hit BALL flew far after it was hit."
banned = ["hit"]

def get_most_common_word(s: str, prohibited: list[str]):
    refined = re.sub('[^a-zA-Z0-9\s]', '',s)
    filtered = list(filter(lambda x: prohibited.count(x) == 0,refined.split()))
    counted  = collections.Counter(map(lambda x: x.lower(),filtered))
    picked = counted.most_common(1)[0][0]
    return list(filter(lambda x: x.lower() == picked,filtered))[0]


get_most_common_word(paragraph, banned)

# in book
def answer(paragraph: str, banned: list[str]) -> str:
    words = [word for word in re.sub(r'[^\w]', ' ', paragraph).lower().split() if word not in banned]
    counts = collections.Counter(words)
    return counts.most_common(1)[0][0]

File: kata-clojure/src/dojo/leetcode/819_most_common_word.clj
(ns dojo.leetcode.819-most-common-word
  (:require [clojure.string :as s]))

(def paragraph "Bob hit a ball, the hit BALL flew far after it was hit.")
(def banned ["hit"])

(defn get-most-common-word
  [p b]
  (let [filtered (filter #(not (contains? (set b) %)) (s/split (s/replace p #"[^a-zA-Z0-9\s]" "") #" "))]
    (key (first (sort-by val > (frequencies (map s/lower-case filtered)))))))

(comment
  (get-most-common-word paragraph banned))
